pkg/kubelet/cm/deviceplugin/device_plugin_handler.go

```diff
@@ -29,79 +29,36 @@ import (
 	"k8s.io/apimachinery/pkg/api/resource"
 	"k8s.io/apimachinery/pkg/util/sets"
 	pluginapi "k8s.io/kubernetes/pkg/kubelet/apis/deviceplugin/v1alpha1"
-	"k8s.io/kubernetes/pkg/kubelet/deviceplugin"
+	kubecontainer "k8s.io/kubernetes/pkg/kubelet/container"
 )
 
-// podDevices represents a list of pod to device Id mappings.
-type containerDevices map[string]sets.String
-type podDevices map[string]containerDevices
-
-func (pdev podDevices) pods() sets.String {
-	ret := sets.NewString()
-	for k := range pdev {
-		ret.Insert(k)
-	}
-	return ret
-}
-
-func (pdev podDevices) insert(podUID, contName string, device string) {
-	if _, exists := pdev[podUID]; !exists {
-		pdev[podUID] = make(containerDevices)
-	}
-	if _, exists := pdev[podUID][contName]; !exists {
-		pdev[podUID][contName] = sets.NewString()
-	}
-	pdev[podUID][contName].Insert(device)
-}
-
-func (pdev podDevices) getDevices(podUID, contName string) sets.String {
-	containers, exists := pdev[podUID]
-	if !exists {
-		return nil
-	}
-	devices, exists := containers[contName]
-	if !exists {
-		return nil
-	}
-	return devices
-}
-
-func (pdev podDevices) delete(pods []string) {
-	for _, uid := range pods {
-		delete(pdev, uid)
-	}
-}
-
-func (pdev podDevices) devices() sets.String {
-	ret := sets.NewString()
-	for _, containerDevices := range pdev {
-		for _, deviceSet := range containerDevices {
-			ret = ret.Union(deviceSet)
-		}
-	}
-	return ret
-}
-
 type DevicePluginHandler interface {
 	// Start starts device plugin registration service.
 	Start() error
 	// Devices returns all of registered devices keyed by resourceName.
 	Devices() map[string][]*pluginapi.Device
 	// Allocate attempts to allocate all of required extended resources for
 	// the input container, issues an Allocate rpc request for each of such
-	// resources, and returns their AllocateResponses on success.
-	Allocate(pod *v1.Pod, container *v1.Container, activePods []*v1.Pod) ([]*pluginapi.AllocateResponse, error)
+	// resources, processes their AllocateResponses, and updates the cached
+	// containerDevices on success.
+	Allocate(pod *v1.Pod, container *v1.Container, activePods []*v1.Pod) error
+	// GetDeviceRunContainerOptions checks whether we have cached containerDevices
+	// for the passed-in <pod, container> and returns its DeviceRunContainerOptions
+	// for the found one. An empty struct is returned in case no cached state is found.
+	GetDeviceRunContainerOptions(pod *v1.Pod, container *v1.Container) DeviceRunContainerOptions
 }
```

Comment:
It would be a bit tricky to see whether we get real containerDevices or not by returning a DeviceRunContainerOptions item. I would prefer to return a reference here.
In addition,  transmission for structs may have impact on performance